crates/net/discv5/src/filter.rs

```diff
@@ -0,0 +1,122 @@
+//! Predicates to constraint peer lookups.
+
+use dashmap::DashSet;
+use derive_more::Constructor;
+use itertools::Itertools;
+
+/// Outcome of applying filtering rules on node record.
+#[derive(Debug, Clone, PartialEq, Eq)]
+pub enum FilterOutcome {
+    /// ENR passes filter rules.
+    Ok,
+    /// ENR doesn't pass filter rules, for the given reason.
+    Ignore {
+        /// Reason for filtering out node record.
+        reason: String,
+    },
+}
+
+impl FilterOutcome {
+    /// Returns `true` for [`FilterOutcome::Ok`].
+    pub fn is_ok(&self) -> bool {
+        matches!(self, FilterOutcome::Ok)
+    }
+}
+
+/// Filter requiring that peers advertise that they belong to some fork of a certain chain.
+#[derive(Debug, Constructor, Clone, Copy, PartialEq, Eq, Hash)]
+pub struct MustIncludeKey {
+    /// Kv-pair key which node record must advertise.
+    key: &'static [u8],
+}
+
+impl MustIncludeKey {
+    /// Returns [`FilterOutcome::Ok`] if [`Enr`](discv5::Enr) contains the configured kv-pair key.
+    pub fn filter(&self, enr: &discv5::Enr) -> FilterOutcome {
+        if enr.get_raw_rlp(self.key).is_none() {
+            return FilterOutcome::Ignore { reason: self.ignore_reason() }
+        }
+        FilterOutcome::Ok
+    }
+
+    fn ignore_reason(&self) -> String {
+        format!("{} fork required", String::from_utf8_lossy(self.key))
+    }
+}
+
+/// Filter requiring that peers not advertise that they belong to some chains.
+#[derive(Debug, Clone, Default)]
+pub struct MustNotIncludeKeys {
+    chains: DashSet<MustIncludeKey>,
```

Comment:
why is this a dashset?

Reply:
I was sending this filter to sigp/discv5, but don't need it anymore